File: DevExpress VCL 23.2.6 Full Source/VCL/ExpressGantt Control/Demos/CBuilder/BaselinesDemo/BaselinesDemoMain.cpp
//---------------------------------------------------------------------------

#include <vcl.h>
#pragma hdrstop

#include "BaselinesDemoMain.h"
//---------------------------------------------------------------------------
#pragma package(smart_init)
#pragma link "cxClasses"
#pragma link "cxContainer"
#pragma link "cxControls"
#pragma link "cxDropDownEdit"
#pragma link "cxEdit"
#pragma link "cxGraphics"
#pragma link "cxLookAndFeelPainters"
#pragma link "cxLookAndFeels"
#pragma link "cxMaskEdit"
#pragma link "cxSpinEdit"
#pragma link "cxTextEdit"
#pragma link "cxButtons"
#pragma link "cxGroupBox"
#pragma link "dxCore"
#pragma link "dxGanttControl"
#pragma link "dxGanttControlAssignments"
#pragma link "dxGanttControlCustomClasses"
#pragma link "dxGanttControlCustomSheet"
#pragma link "dxGanttControlResources"
#pragma link "dxGanttControlTasks"
#pragma link "dxGanttControlViewChart"
#pragma link "dxGanttControlViewResourceSheet"
#pragma link "dxGanttControlViewTimeline"
#pragma link "dxLayoutContainer"
#pragma link "dxLayoutControl"
#pragma link "dxLayoutControlAdapters"
#pragma link "dxLayoutcxEditAdapters"
#pragma link "dxLayoutLookAndFeels"
#pragma link "cxImageList"
#pragma link "dxShellDialogs"
#pragma resource "*.dfm"
TBaselinesDemoMainForm *BaselinesDemoMainForm;
//---------------------------------------------------------------------------
__fastcall TBaselinesDemoMainForm::TBaselinesDemoMainForm(TComponent* Owner)
	: TDemoBasicMainForm(Owner)
{
  GanttControl->DataModel->LoadFromFile("..\\..\\Data\\SoftDev-Baselines.xml");
}
//---------------------------------------------------------------------------
void __fastcall TBaselinesDemoMainForm::GanttControlDataModelLoaded(TObject *Sender)
{
  UpdateBaselineList();
  cbBaseline->ItemIndex = 0;
  cbSetBaseline->ItemIndex = 0;
}
//---------------------------------------------------------------------------

void __fastcall TBaselinesDemoMainForm::GanttControlBaselineChanged(TObject *Sender,
		  TdxGanttControlDataModelBaseline *ABaseline)
{
  UpdateBaselineList();
}
//---------------------------------------------------------------------------

void __fastcall TBaselinesDemoMainForm::UpdateBaselineList()
{
  FLockCount++;
  cbBaseline->ActiveProperties->BeginUpdate();
  cbSetBaseline->ActiveProperties->BeginUpdate();
  try {
  	TdxGanttControlDataModelBaseline *ABaseline;
	if (GanttControl->ViewChart->BaselineNumber < 0)
	  ABaseline = NULL;
	else
	  ABaseline = GanttControl->DataModel->Baselines->Find(GanttControl->ViewChart->BaselineNumber);
	int ASetBaselineIndex = cbSetBaseline->ItemIndex;
	cbBaseline->ActiveProperties->Items->Clear();
	cbSetBaseline->ActiveProperties->Items->Clear();
	cbSetBaseline->ActiveProperties->Items->Add("Baseline");
	for (int I = 1; I <= 10; I++)
	  cbSetBaseline->ActiveProperties->Items->Add("Baseline " + IntToStr(I));
	cbBaseline->ActiveProperties->Items->Add("None");
	for (int I = 0; I < GanttControl->DataModel->Baselines->Count; I++) {
	  UnicodeString AText = BaselineToText(GanttControl->DataModel->Baselines->Items[I]);
	  cbBaseline->ActiveProperties->Items->AddObject(AText, GanttControl->DataModel->Baselines->Items[I]);
	  if (GanttControl->DataModel->Baselines->Items[I]->Number <= 10) {
		cbSetBaseline->ActiveProperties->Items->Strings[GanttControl->DataModel->Baselines->Items[I]->Number] = AText;
		cbSetBaseline->ActiveProperties->Items->Objects[GanttControl->DataModel->Baselines->Items[I]->Number] = GanttControl->DataModel->Baselines->Items[I];
	  }
	}
	int I = cbBaseline->ActiveProperties->Items->IndexOfObject(ABaseline);
	cbBaseline->ItemIndex = Max(0, I);
	cbSetBaseline->ItemIndex = ASetBaselineIndex;
  } __finally {
	cbSetBaseline->ActiveProperties->EndUpdate();
	cbBaseline->ActiveProperties->EndUpdate();
	FLockCount--;
  }
}
//---------------------------------------------------------------------------

UnicodeString __fastcall TBaselinesDemoMainForm::BaselineToText(TdxGanttControlDataModelBaseline *ABaseline)
{
  if (ABaseline->Description != "")
	return ABaseline->Description;
  if (ABaseline->Number == 0)
	return "Baseline (Last saved on " + DateTimeToStr(ABaseline->Created) + ")";
  else
	return "Baseline "+ IntToStr(ABaseline->Number) + " (Last saved on " + DateTimeToStr(ABaseline->Created) + ")";
}
//---------------------------------------------------------------------------

void __fastcall TBaselinesDemoMainForm::GanttControlViewChartBaselineNumberChanged(TObject *Sender)

{
  if (GanttControl->ViewChart->BaselineNumber < 0)
	cbBaseline->ItemIndex = 0;
  else {
	for (int I = 1; I < cbBaseline->ActiveProperties->Items->Count; I++)
	  if (((TdxGanttControlDataModelBaseline*)(cbBaseline->ActiveProperties->Items->Objects[I]))->Number == GanttControl->ViewChart->BaselineNumber) {
		cbBaseline->ItemIndex = I;
		break;
	  }
  }
}
//---------------------------------------------------------------------------

void __fastcall TBaselinesDemoMainForm::cbBaselinePropertiesEditValueChanged(TObject *Sender)

{
  if (FLockCount > 0)
	return;
  if ((cbBaseline->ItemIndex == -1) || (cbBaseline->ActiveProperties->Items->Objects[cbBaseline->ItemIndex] == NULL))
	GanttControl->ViewChart->HideBaselines();
  else
	GanttControl->ViewChart->ShowBaselines(((TdxGanttControlDataModelBaseline*)(cbBaseline->ActiveProperties->Items->Objects[cbBaseline->ItemIndex]))->Number);
}
//---------------------------------------------------------------------------

void __fastcall TBaselinesDemoMainForm::cxButton3Click(TObject *Sender)
{
  GanttControl->ViewChart->SetBaselines(cbSetBaseline->ItemIndex, lcbiSelectedTasks->Checked);
}
//---------------------------------------------------------------------------

void __fastcall TBaselinesDemoMainForm::cxButton4Click(TObject *Sender)
{
  GanttControl->ViewChart->ClearBaselines(cbSetBaseline->ItemIndex, lcbiSelectedTasks->Checked);
}
//---------------------------------------------------------------------------

File: DevExpress VCL 23.2.6 Full Source/VCL/ExpressGantt Control/Demos/Delphi/BaselinesDemo/BaselinesDemoMain.pas
unit BaselinesDemoMain;

interface

{$I cxVer.inc}

uses
  Windows, Messages, SysUtils, Classes, Graphics, Controls, Forms,
  Dialogs, Menus, ComCtrls, StdCtrls,
  dxCore, cxContainer, cxEdit, cxTextEdit, cxDropDownEdit, cxMaskEdit, cxSpinEdit,
  cxGraphics, cxClasses, cxControls,
{$IFDEF EXPRESSSKINS}
  dxSkinsdxRibbonPainter,
  dxSkinsdxBarPainter,
{$ENDIF}
  cxLookAndFeels, cxLookAndFeelPainters, dxLayoutControl,
  dxLayoutContainer, dxLayoutLookAndFeels, dxLayoutcxEditAdapters, DemoBasicMain,
  dxGanttControlCustomClasses, dxGanttControl, dxGanttControlCustomSheet, dxGanttControlViewChart,
  dxGanttControlViewResourceSheet, dxGanttControlViewTimeline, dxGanttControlTasks, dxGanttControlAssignments,
  dxGanttControlDataModel, dxGanttControlResources, dxLayoutControlAdapters,
  ActnList, ImgList, cxImageList, dxShellDialogs, cxButtons, cxGroupBox;

type
  TBaselinesDemoMainForm = class(TDemoBasicMainForm)
    dxLayoutGroup9: TdxLayoutGroup;
    cbBaseline: TcxComboBox;
    dxLayoutItem8: TdxLayoutItem;
    dxLayoutItem9: TdxLayoutItem;
    cbSetBaseline: TcxComboBox;
    lcbiSelectedTasks: TdxLayoutCheckBoxItem;
    cxButton3: TcxButton;
    dxLayoutItem10: TdxLayoutItem;
    cxButton4: TcxButton;
    dxLayoutItem11: TdxLayoutItem;
    dxLayoutAutoCreatedGroup1: TdxLayoutAutoCreatedGroup;
    procedure GanttControlDataModelLoaded(Sender: TObject);
    procedure GanttControlViewChartBaselineNumberChanged(Sender: TObject);
    procedure GanttControlBaselineChanged(Sender: TObject; ABaseline: TdxGanttControlDataModelBaseline);
    procedure FormCreate(Sender: TObject);
    procedure cbBaselinePropertiesEditValueChanged(Sender: TObject);
    procedure cxButton3Click(Sender: TObject);
    procedure cxButton4Click(Sender: TObject);
  private
    FLockCount: Integer;
    function BaselineToText(ABaseline: TdxGanttControlDataModelBaseline): string;
    procedure UpdateBaselineList;
  end;

var
  BaselinesDemoMainForm: TBaselinesDemoMainForm;

implementation

{$R *.dfm}

uses
  Math,
  dxCoreClasses,
  dxGanttControlCustomDataModel;

{ TBaselinesDemoMainForm }

function TBaselinesDemoMainForm.BaselineToText(
  ABaseline: TdxGanttControlDataModelBaseline): string;
begin
  if ABaseline.Description <> '' then
    Result := ABaseline.Description
  else if ABaseline.Number = 0 then
    Result := Format('Baseline (Last saved on %s)', [DateTimeToStr(ABaseline.Created)])
  else
    Result := Format('Baseline %d (Last saved on %s)', [ABaseline.Number, DateTimeToStr(ABaseline.Created)])
end;

procedure TBaselinesDemoMainForm.cbBaselinePropertiesEditValueChanged(
  Sender: TObject);
begin
  if FLockCount > 0 then
    Exit;
  if (cbBaseline.ItemIndex = -1) or (cbBaseline.ActiveProperties.Items.Objects[cbBaseline.ItemIndex] = nil) then
    GanttControl.ViewChart.HideBaselines
  else
    GanttControl.ViewChart.ShowBaselines(TdxGanttControlDataModelBaseline(cbBaseline.ActiveProperties.Items.Objects[cbBaseline.ItemIndex]).Number);
end;

procedure TBaselinesDemoMainForm.cxButton3Click(Sender: TObject);
begin
  GanttControl.ViewChart.SetBaselines(cbSetBaseline.ItemIndex, lcbiSelectedTasks.Checked);
end;

procedure TBaselinesDemoMainForm.cxButton4Click(Sender: TObject);
begin
  GanttControl.ViewChart.ClearBaselines(cbSetBaseline.ItemIndex, lcbiSelectedTasks.Checked);
end;

procedure TBaselinesDemoMainForm.FormCreate(Sender: TObject);
begin
  inherited;
  GanttControl.DataModel.LoadFromFile('..\..\Data\SoftDev-Baselines.xml');
end;

procedure TBaselinesDemoMainForm.GanttControlDataModelLoaded(Sender: TObject);
begin
  inherited;
  UpdateBaselineList;
  cbBaseline.ItemIndex := 0;
  cbSetBaseline.ItemIndex := 0;
end;

procedure TBaselinesDemoMainForm.GanttControlViewChartBaselineNumberChanged(Sender: TObject);
var
  I: Integer;
begin
  if GanttControl.ViewChart.BaselineNumber < 0 then
    cbBaseline.ItemIndex := 0
  else
  begin
    for I := 1 to cbBaseline.ActiveProperties.Items.Count - 1 do
      if TdxGanttControlDataModelBaseline(cbBaseline.ActiveProperties.Items.Objects[I]).Number = GanttControl.ViewChart.BaselineNumber then
      begin
        cbBaseline.ItemIndex := I;
        Break;
      end;
  end;
end;

procedure TBaselinesDemoMainForm.GanttControlBaselineChanged(Sender: TObject; ABaseline: TdxGanttControlDataModelBaseline);
begin
  UpdateBaselineList;
end;

procedure TBaselinesDemoMainForm.UpdateBaselineList;
var
  I: Integer;
  AText: string;
  ABaseline: TdxGanttControlDataModelBaseline;
  ASetBaselineIndex: Integer;
begin
  Inc(FLockCount);
  cbBaseline.ActiveProperties.BeginUpdate;
  cbSetBaseline.ActiveProperties.BeginUpdate;
  try
    if GanttControl.ViewChart.BaselineNumber < 0 then
      ABaseline := nil
    else
      ABaseline := GanttControl.DataModel.Baselines.Find(GanttControl.ViewChart.BaselineNumber);
    ASetBaselineIndex := cbSetBaseline.ItemIndex;
    cbBaseline.ActiveProperties.Items.Clear;
    cbSetBaseline.ActiveProperties.Items.Clear;
    cbSetBaseline.ActiveProperties.Items.Add('Baseline');
    for I := 1 to 10 do
      cbSetBaseline.ActiveProperties.Items.Add(Format('Baseline %d', [I]));
    cbBaseline.ActiveProperties.Items.Add('None');
    for I := 0 to GanttControl.DataModel.Baselines.Count - 1 do
    begin
      AText := BaselineToText(GanttControl.DataModel.Baselines[I]);
      cbBaseline.ActiveProperties.Items.AddObject(AText, GanttControl.DataModel.Baselines[I]);
      if GanttControl.DataModel.Baselines[I].Number <= 10 then
      begin
        cbSetBaseline.ActiveProperties.Items[GanttControl.DataModel.Baselines[I].Number] := AText;
        cbSetBaseline.ActiveProperties.Items.Objects[GanttControl.DataModel.Baselines[I].Number] := GanttControl.DataModel.Baselines[I];
      end;
    end;
    I := cbBaseline.ActiveProperties.Items.IndexOfObject(ABaseline);
    cbBaseline.ItemIndex := Max(0, I);
    cbSetBaseline.ItemIndex := ASetBaselineIndex;
  finally
    cbSetBaseline.ActiveProperties.EndUpdate;
    cbBaseline.ActiveProperties.EndUpdate;
    Dec(FLockCount);
  end;
end;

end.
